### PR TITLE
Uploading T1w and T2w pediatric MRI templates generated from philadelphia-pediatric dataset

### LINKED ISSUE #6: Create first version of template and upload as release asset
That way we can download it via SCT.

Also see: https://github.com/neuropoly/data-management/issues/248#issuecomment-1731691329

### LINKED ISSUE #7: Document how to create the pediatric template
Edit the README to add the following info:
- code used to generate the template: https://github.com/neuropoly/template, indicate the version used
- the pediatric-specific files used (eg: configuration.json, data, version of the data, etc.)

### PR DESCRIPTION
**Pull request by NadiaBlostein**
Fixes #6, Fixes #7 

Screen Shot 2023-09-29 at 3 36 11 PM

Screen Shot 2023-09-29 at 3 36 23 PM

### CODE REVIEW

**Comment by jcohenadad** [comment]
@NadiaBlostein Thank you for this. My suggestion was to upload the templates as release assets, not directly in the git history. The JSON files can stay though. If you don't know how to do this I can take care of it.

**Comment by jcohenadad** [comment]
I also noticed a few issues:
- the nz (dim3) between T1w (dim3=687) and T2w (dim3=680) is not the same
- the sform between T1w and T2w is not the same

<details><summary>pediatric-template_T1w.nii.gz</summary>

```console
julien-macbook:~/Desktop $ fslhd pediatric-template_T1w.nii.gz 
filename	pediatric-template_T1w.nii.gz

sizeof_hdr	348
data_type	FLOAT32
dim0		3
dim1		201
dim2		201
dim3		687
dim4		0
dim5		0
dim6		0
dim7		0
vox_units	mm
time_units	Unknown
datatype	16
nbyper		4
bitpix		32
pixdim0		0.000000
pixdim1		0.500000
pixdim2		0.500000
pixdim3		0.500000
pixdim4		0.000000
pixdim5		0.000000
pixdim6		0.000000
pixdim7		0.000000
vox_offset	352
cal_max		0.000000
cal_min		0.000000
scl_slope	0.000000
scl_inter	0.000000
phase_dim	0
freq_dim	0
slice_dim	0
slice_name	Unknown
slice_code	0
slice_start	0
slice_end	0
slice_duration	0.000000
toffset		0.000000
intent		Unknown
intent_code	0
intent_name	
intent_p1	0.000000
intent_p2	0.000000
intent_p3	0.000000
qform_name	Unknown
qform_code	0
qto_xyz:1	0.500000 0.000000 0.000000 0.000000 
qto_xyz:2	0.000000 0.500000 0.000000 0.000000 
qto_xyz:3	0.000000 0.000000 0.500000 0.000000 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Left-to-Right
qform_yorient	Posterior-to-Anterior
qform_zorient	Inferior-to-Superior
sform_name	Unknown
sform_code	1
sto_xyz:1	-0.500000 0.000000 0.000000 50.250000 
sto_xyz:2	-0.000000 0.500000 0.000000 -50.250000 
sto_xyz:3	-0.000000 0.000000 0.500000 -336.818176 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Right-to-Left
sform_yorient	Posterior-to-Anterior
sform_zorient	Inferior-to-Superior
file_type	NIFTI-1+
file_code	1
descrip		mnc2nii avg.007.mnc
aux_file	
```

</details>

<details><summary>Terminal output</summary>

```console
julien-macbook:~/Desktop $ fslhd pediatric-template_T2w.nii.gz 
filename	pediatric-template_T2w.nii.gz

sizeof_hdr	348
data_type	FLOAT32
dim0		3
dim1		201
dim2		201
dim3		680
dim4		0
dim5		0
dim6		0
dim7		0
vox_units	mm
time_units	Unknown
datatype	16
nbyper		4
bitpix		32
pixdim0		0.000000
pixdim1		0.500000
pixdim2		0.500000
pixdim3		0.500000
pixdim4		0.000000
pixdim5		0.000000
pixdim6		0.000000
pixdim7		0.000000
vox_offset	352
cal_max		0.000000
cal_min		0.000000
scl_slope	0.000000
scl_inter	0.000000
phase_dim	0
freq_dim	0
slice_dim	0
slice_name	Unknown
slice_code	0
slice_start	0
slice_end	0
slice_duration	0.000000
toffset		0.000000
intent		Unknown
intent_code	0
intent_name	
intent_p1	0.000000
intent_p2	0.000000
intent_p3	0.000000
qform_name	Unknown
qform_code	0
qto_xyz:1	0.500000 0.000000 0.000000 0.000000 
qto_xyz:2	0.000000 0.500000 0.000000 0.000000 
qto_xyz:3	0.000000 0.000000 0.500000 0.000000 
qto_xyz:4	0.000000 0.000000 0.000000 1.000000 
qform_xorient	Left-to-Right
qform_yorient	Posterior-to-Anterior
qform_zorient	Inferior-to-Superior
sform_name	Unknown
sform_code	1
sto_xyz:1	-0.500000 0.000000 0.000000 50.250000 
sto_xyz:2	-0.000000 0.500000 0.000000 -50.250000 
sto_xyz:3	-0.000000 0.000000 0.500000 -333.203949 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Right-to-Left
sform_yorient	Posterior-to-Anterior
sform_zorient	Inferior-to-Superior
file_type	NIFTI-1+
file_code	1
descrip		mnc2nii avg.007.mnc
aux_file	
```

</details>

**Comment by NadiaBlostein** [comment]
Hi @jcohenadad,

I did not know that the two templates had to be built in the same space. Below is an overview of what I understood, to make sure we are on the same page:

## Summary of the template-generation process (what I undersood)

**1.** An [initial template space is generated](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L389) based on the average length of the intervertebral levels of the subjects in the dataset of interest. It gets saved downstream as a file called `template_mask.mnc`.

**2.** The [straightening and normalization method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) straightens all of the spinal cord images of the subjects and brings them into the same space as `template_mask.mnc`.

**3.** An iterative vowel-wise inter-subject averaging begins.

## Making sure we are on the same page: details & questions

**1.** An [initial template space is generated](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L389) based on the average length of the intervertebral levels of the subjects in the dataset of interest. It gets saved downstream as a file called `template_mask.mnc`. **CHECKPOINT 1**:
* For the T1w images of the pediatric dataset, `template_mask.mnc` has _xyz_ dimensions _201 x 201  x 692_.
* For the T2w images of the pediatric dataset, `template_mask.mnc` has _xyz_ dimensions _201 x 201 x 687_.
* In the pediatric dataset, the dimensions of the T1w and T2w images for the same subject are usually different (_e.g. for sub-101, the recomposed T1w image is 288 x 582 x 191 whereas the recomposed T2w image is 328 x 580 x 63_). 
* **Therefore, it DOES make sense that `template_mask.mnc` for the T1w images and `template_mask.mnc` for the T2w images would have different dimensions and sforms.** Is this correct?

**2.** The [straightening and normalization method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) straightens all of the spinal cord images of the subjects and brings them into the same space as `template_mask.mnc`. **CHECKPOINT 2 – where I think there is a bug.**
* For the T1w images of the pediatric dataset, this outputs straightened and normalized images with dimensions _201 x 201 x 687_ for 5 subjects and dimensions _201 x 201 x 692_ for 13 subjects. **This is the step where something fishy is going on because we would expect 18 images with dimensions  _201 x 201 x 692_**.
* For the T2w images of the pediatric dataset, this outputs straightened and normalized images with dimensions _201 x 201 x 680_ for 5 subjects and dimensions _201 x 201 x 687_ for 13 subjects. **This is the step where something fishy is going on because we would expect 18 images with dimensions  _201 x 201 x 687_**.

**3.** The iterative vowel-wise inter-subject averaging begins.
* For the T1w images of the pediatric dataset, this outputs a file with dimensions _201 x 201 x 687_ (smallest dimensions of straightened and normalized T1w images, see Step 2). 
* For the T2w images of the pediatric dataset, this outputs a file with dimensions _201 x 201 x 680_ (smallest dimensions of straightened and normalized T2w images, see Step 2). 
* **Step 2 therefore needs to be fixed.**

## Moving forward

1) Does it make sense now why the T1w and T2w templates would have different dimensions?

2) I will double check that the `straighten_all_subjects` [method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) of the template pipeline is where something is failing and I will open an issue on on [Neuropoly/Template](https://github.com/neuropoly/template/tree/master).

**Comment by NadiaBlostein** [comment]
@jcohenadad I just re-uploaded the files as release assets.

**Comment by jcohenadad** [comment]
Thank you for the thorough explanation @NadiaBlostein 

> Does it make sense now why the T1w and T2w templates would have different dimensions?

Yes it does make sense. So I'm wondering how @benjamindeleener managed to get the same voxel and world space coordinates. @benjamindeleener can you please comment?

> I will double check that the straighten_all_subjects [method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) of the template pipeline is where something is failing and I will open an issue on on [Neuropoly/Template](https://github.com/neuropoly/template/tree/master).

Excellent-- thank you

**Comment by benjamindeleener** [comment]
I don't think they should necessarily have different sizes as both template represent and are generated from the same subjects. The way I ensured both T1w and T2w have the same size was to use only one initial template. For example, you generate the initial template using T1w along with all the labels (centerline and disks locations) and you straighten the T1w **and** the T2w images to this initial template. You should therefore have all the images into the same space.

File: README.md [hunk 1]
```diff
@@ -7,3 +7,194 @@ Repository for generating pediatric template
...
+
+## Regenerating T1w and T2w Templates
+Step-by-step walk-through on how to generate T1w and T2w pediatric templates from philadelphia-pediatric dataset.
+All code below should be run in your command-line (bash) on CCDB. Each step begins from your **working directory**: `/PATH/TO/pediatric-template`.
+
+### 1. Get data from git annex
+```
+git clone git@data.neuro.polymtl.ca:datasets/philadelphia-pediatric
+cd philadelphia-pediatric
+git checkout -b nb/add-labels
```
Comment (rohanbanerjee) [inline README.md:26]: This branch is no longer required. Everything has been merged into the master branch. See -- https://github.com/neuropoly/data-management/issues/248#issuecomment-1745287547

**Comment by NadiaBlostein** [comment]
Hi @jcohenadad,

I just re-uploaded the templates – everything should be correct now, see [latest release](https://github.com/spinalcordtoolbox/pediatric-template/tree/r20231010). Explanation below.

## Summary of template-generation (repeated from [comment above](https://github.com/spinalcordtoolbox/pediatric-template/pull/9#issuecomment-1742664210))
1. An [initial template space is generated](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L389) based on the average length of the intervertebral levels of the subjects in the dataset of interest. It gets saved downstream as a file called `template_mask.mnc`.

2. The [straightening and normalization method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) straightens all of the spinal cord images of the subjects and brings them into the same space as `template_mask.mnc`.

3. An iterative vowel-wise inter-subject averaging begins.

## Current templates + [previous error](https://github.com/spinalcordtoolbox/pediatric-template/pull/9#issuecomment-1742664210) explained

1. An [initial template space is generated](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L389) based on the average length of the intervertebral levels of the subjects in the dataset of interest. It gets saved downstream as a file called `template_mask.mnc`. CHECKPOINT 1:
   * For the T1w images of the pediatric dataset, `template_mask.mnc` has xyz dimensions **201 x 201 x 688**.
   * For the T2w images of the pediatric dataset,  `template_mask.mnc` has xyz dimensions **201 x 201 x 680**.

2. The [straightening and normalization method](https://github.com/neuropoly/template/blob/a7915f4ccfa075a5d31f4ea84bb9761d42710e9e/preprocess_normalize.py#L463C1-L463C1) straightens all of the spinal cord images of the subjects and brings them into the same space as `template_mask.mnc`.
   * For the T1w images of the pediatric dataset, all of the straightened images have xyz dimensions **201 x 201 x 688**.
   * For the T2w images of the pediatric dataset, all of the straightened images have xyz dimensions **201 x 201 x 680**.
   * ERROR EXPLAINED: The reason why this was not the case prior was due to an error on my part: the straightening and normalization has to be performed on ALL of the subjects at once such that `template_mask.mnc` is generated correctly. I originally performed the straightening and normalization in batches, meaning that the dimensions of `template_mask.mnc` were not based on an average across all subejcts. Is this clear?

3. The iterative vowel-wise inter-subject averaging begins.
   * For the T1w images of the pediatric dataset, this outputs a file (saved as `pediatric-template_T1w.nii.gz` after 8 iterations) with dimensions **201 x 201 x 688**.
   * For the T2w images of the pediatric dataset, this outputs a file (saved as `pediatric-template_T2w.nii.gz` after 8 iterations) with dimensions **201 x 201 x 680**.

## Moving forward
Is this clear and correct?

Thank you

**Comment by jcohenadad** [comment]
Thank you @NadiaBlostein, your explanation is clear, but I think it does not address the solution that @benjamindeleener mentioned, ie: using the _same_ target space for both contrasts (you still generate each template in its own image space).

But since the release is done, we can merge this PR and address this issue in a subsequent PR. I've opened an issue: https://github.com/spinalcordtoolbox/pediatric-template/issues/10

I've also added more context in the release text (under 'details'): https://github.com/spinalcordtoolbox/pediatric-template/releases/tag/r20231010

**Pull request merged**